### PR TITLE
#167730648 Notification for approved request

### PR DESCRIPTION
**Pull request by chuksjoe**
#### What does this PR do?
Setup Notification for approved travel request

#### Description of Task to be completed?

- setup pusher for sending out a notification
- send out a notification to a specific requester when his/her travel request is approved
- create a notification table
- create a route for getting all notification for a user
- create route for getting a specific notification
- test the notification with an HTML file
- test the route endpoints for retrieving notifications

#### How should this be manually tested?

- clone this repo into your local machine
- create a `.env` file in the root folder and copy the variables in the `.env.sample` into it, with their appropriate values
- run `npm install` then `npm run dev`

#### What are the relevant pivotal tracker stories?

[#167730648](https://www.pivotaltracker.com/story/show/167730648)

#### Screenshots (if appropriate)
Travel Approval Notification Screenshot

#### Questions:

### CODE REVIEW

File: src/controllers/travelControllers.js [hunk 1]
```diff
@@ -110,13 +111,31 @@ const getUserTravelStatus = async (req, res) => {
...
+    const notificationObj = await createNotification({
+      recipient_id: updatedTravel.user_id,
+      travel_id: updatedTravel.id,
+      message:
+        `Your travel request created on ${updatedTravel.createdAt} with ${updatedTravel.origin} as your origin has been approved.`,
+    });
+    const notificationMessage = notificationObj.dataValues;
+
+    delete notificationMessage.recipient_id;
+    notificationMessage.title = 'Approved Travel Request.';
```
Comment (Kaytbode) [inline src/controllers/travelControllers.js:128]: @chuksjoe Good work

why not put the messages **(notificationMessage.title  and the very long one)** in the` utils/utilsmessages` folder?

**Review by Kaytbode — COMMENTED** [review]
@chuksjoe  I am fine with what you have done, however just need some clarifications.

why not put the messages (notificationMessage.title and _the very long one_) in the utils/utilsmessages folder?

**Comment by chuksjoe** [comment]
> @chuksjoe I am fine with what you have done, however just need some clarifications.
> 
> why not put the messages (notificationMessage.title and _the very long one_) in the utils/utilsmessages folder?

Thanks bro @Kaytbode. I have implemented it.

**Review by Kaytbode — APPROVED** [review]

**Review by Marshalheri — COMMENTED** [review]
Great job team mate.
I just want to point out that in the model for the notification, i think it should have a relationship with the travel_request table for the travel_id column. Please look into that, and do give clarity if it is not needed.
Thanks.

**Comment by chuksjoe** [comment]
> Great job teammate.
> I just want to point out that in the model for the notification, I think it should have a relationship with the travel_request table for the travel_id column. Please look into that, and do give clarity if it is not needed.
> Thanks.

Good morning bro @Marshalheri, thanks for taking out time to review my PR.
The notification is tied to a user who its sent to. The travel ID in the notification table is just for reference to the travel request for which the notification is sent.
The notification is just like a message sent to the requester or manager as the case may be.

**Review by danoseun — COMMENTED** [review]
Nice job bro.How did you test this functionality?

**Comment by Marshalheri** [comment]
> > Great job teammate.
> > I just want to point out that in the model for the notification, I think it should have a relationship with the travel_request table for the travel_id column. Please look into that, and do give clarity if it is not needed.
> > Thanks.
> 
> Good morning bro @Marshalheri, thanks for taking out time to review my PR.
> The notification is tied to a user who its sent to. The travel ID in the notification table is just for reference to the travel request for which the notification is sent.
> The notification is just like a message sent to the requester or manager as the case may be.

Ok bro. Thanks for the explanation, but if thats the case then i guess i can actually just input any travel_id and it probably might not be one that exist on the db or is even related to that notification.
Also please add **encrypted: true** to the config object for the pusher class. I will be adding it from my own side so that our file can be in sync.
Thanks.

**Comment by chuksjoe** [comment]
> Nice job bro. How did you test this functionality?

I tested it with the index.html file in the root directory. Though, I wrote tests for the routes that get the notifications from the database.

**Comment by chuksjoe** [comment]
> > > Great job teammate.
> > > I just want to point out that in the model for the notification, I think it should have a relationship with the travel_request table for the travel_id column. Please look into that, and do give clarity if it is not needed.
> > > Thanks.
> > 
> > 
> > Good morning bro @Marshalheri, thanks for taking out time to review my PR.
> > The notification is tied to a user who its sent to. The travel ID in the notification table is just for reference to the travel request for which the notification is sent.
> > The notification is just like a message sent to the requester or manager as the case may be.
> 
> Ok bro. Thanks for the explanation, but if that's the case then I guess I can actually just input any travel_id and it probably might not be one that exists on the DB or is even related to that notification.
> Also please add **encrypted: true** to the config object for the pusher class. I will be adding it from my own side so that our file can be in sync.
> Thanks.

It uses the travel_id of the travel that was approved which is gotten from the database, as such it cannot be invalid. But the notification is more related to the person it's been sent to which is why the recipient_id is made the foreign key. Also, I have added the encrypted property to the pusher.

Thanks for your observation.

**Comment by Marshalheri** [comment]
> > > > Great job teammate.
> > > > I just want to point out that in the model for the notification, I think it should have a relationship with the travel_request table for the travel_id column. Please look into that, and do give clarity if it is not needed.
> > > > Thanks.
> > > 
> > > 
> > > Good morning bro @Marshalheri, thanks for taking out time to review my PR.
> > > The notification is tied to a user who its sent to. The travel ID in the notification table is just for reference to the travel request for which the notification is sent.
> > > The notification is just like a message sent to the requester or manager as the case may be.
> > 
> > 
> > Ok bro. Thanks for the explanation, but if that's the case then I guess I can actually just input any travel_id and it probably might not be one that exists on the DB or is even related to that notification.
> > Also please add **encrypted: true** to the config object for the pusher class. I will be adding it from my own side so that our file can be in sync.
> > Thanks.
> 
> It uses the travel_id of the travel that was approved which is gotten from the database, as such it cannot be invalid. But the notification is more related to the person it's been sent to which is why the recipient_id is made the foreign key. Also, I have added the encrypted property to the pusher.
> 
> Thanks for your observation.

Ok thanks i will approve now..

**Review by Marshalheri — APPROVED** [review]

**Pull request merged**